packages/server/src/controllers/roadmaps/filter.ts

```diff
@@ -1,23 +1,79 @@
+import { z } from "zod"; // Add this import
 import database from "../../database";
-
-// utils
 import logger from "../../utils/logger";
 import error from "../../errorResponse.json";
 
-export async function filter(_, res) {
+// Move the schema from route to controller
+const querySchema = z.object({
+  first: z.coerce.number().min(1).max(100).default(20),
+  after: z.string().uuid().optional()
+});
+
+export async function filter(req, res) {
   try {
-    const roadmaps = await database
+    // Validate query parameters at the start of controller
+    const { first, after } = querySchema.parse(req.query);
+
+    // Build cursor query
+    let query = database
       .select("id", "name", "url", "color", "display", "index")
       .from("roadmaps")
-      .orderBy("index", "asc");
+      .orderBy("id", "asc")
+      .limit(first + 1);
 
-    res.status(200).send({ roadmaps });
-  } catch (err) {
-    logger.error({
-      message: err,
+    if (after) {
+      query = query.where("id", ">", after);
+    }
+
+    const rows = await query;
+
+    // Determine pagination flags
+    const hasNextPage = rows.length > first;
+    const data = hasNextPage ? rows.slice(0, first) : rows;
+
+    // Get total count (only on first page for performance)
+    let totalCount: number | null = null;
+    let totalPages: number | null = null;
+
+    if (!after) {
+      const countResult = await database('roadmaps').count('* as count');
+      totalCount = Number(countResult[0].count);
+      totalPages = Math.ceil(totalCount / first);
+    }
+
+    // Calculate current page
+    let currentPage = 1;
+    if (after) {
+      const afterCountResult = await database('roadmaps')
+        .where('id', '<=', after)
+        .count('* as count');
+      const afterCount = Number(afterCountResult[0].count);
+      currentPage = Math.floor(afterCount / first) + 1;
+    }
+
+    // Response with consistent format
+    res.status(200).json({
+      data,
```

Comment:
@kris70lesgo Can you also add `roadmaps` property? This is just for compatibility.

To avoid breaking changes with any existing integrations.